### PR TITLE
Workaround GraphNodeRef static_assert that does not compile with IBM XL

### PR DESCRIPTION
**Pull request by dalg24**
Reported by @ndellingwood and reproduced on Summit
```
core/src/Kokkos_GraphNode.hpp:75:59: error: template argument for template template parameter must be a class template or type
      alias template
          Kokkos::Impl::is_specialization_of<Predecessor, GraphNodeRef>::value,
                                                          ^
```

Updated `<Kokkos_Macros.hpp>` to detect the use of `xlclang++` and dropped use of legacy IBM XL-based built-in functions in `<Kokkos_BitOps.hpp>`.

### CODE REVIEW

**Comment by dalg24** [comment]
Error with MSVC seems unrelated
https://ci.appveyor.com/project/dalg24/kokkos/builds/36325633?fullLog=true#L6400

**Review by DavidPoliakoff — APPROVED** [review]
Oof. Rough one, but a good fix for a 3.3 timeline

**Comment by DavidPoliakoff** [comment]
(obviously, need the Appveyor build to pass, whatever that's about)

```
C:\projects\source\core\src\impl/Kokkos_TaskNode.hpp(422,1): error C2325: 'Kokkos::Impl::TaskNode<TaskQueueTraits>': unexpected type to the right of '.~': expected 'Kokkos::Impl::TaskNode<TaskQueueTraits> *' [C:\projects\source\build\core\perf_test\KokkosCore_PerformanceTest_TaskDag.vcxproj]
          with
          [
              TaskQueueTraits=Kokkos::Impl::TaskQueueTraitsLockBased
          ]
C:\projects\source\core\src\impl/Kokkos_TaskQueueMemoryManager.hpp(131): message : see reference to function template instantiation 'Kokkos::Impl::AggregateTask<Kokkos::Impl::TaskQueueTraitsLockBased,Kokkos::Impl::EmptyTaskSchedulingInfo>::AggregateTask<_Ty,int,Kokkos::Impl::TaskQueueMemoryManager<ExecSpace,MemorySpace,MemoryPool>::allocation_size_type&>(int32_t,_Ty &&,int &&,Kokkos::Impl::TaskQueueMemoryManager<ExecSpace,MemorySpace,MemoryPool>::allocation_size_type &)' being compiled [C:\projects\source\build\core\perf_test\KokkosCore_PerformanceTest_TaskDag.vcxproj]
          with
          [
```

**Comment by dalg24** [comment]
Same error with #3591

**Comment by dalg24** [comment]
```
core/src/impl/Kokkos_BitOps.hpp(66): error: identifier "__cntlz4" is undefined
core/src/impl/Kokkos_BitOps.hpp(96): error: identifier "__cnttz4" is undefined
core/src/impl/Kokkos_BitOps.hpp(118): error: identifier "__cnttz4" is undefined
core/src/impl/Kokkos_BitOps.hpp(141): error: identifier "__popcnt4" is undefined

**Comment by dalg24** [comment]
We decided to drop the use of built-in functions supported by the legacy IBM XL C++ compiler.  In principle, we could look for the built-in functions supported by xlclang++ but we chose not to do so yet to allow hypothetical builds with the legacy XL-based compiler.

**Review by Rombur — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -154,13 +154,16 @@
 #define KOKKOS_COMPILER_IBM __IBMCPP__
 #elif defined(__IBMC__)
 #define KOKKOS_COMPILER_IBM __IBMC__
+#elif defined(__ibmxl_vrm__)  // xlclang++
```
Comment (ndellingwood) [inline core/src/Kokkos_Macros.hpp:157]: These changes did not work for me (cuda/10.1 builds failed), I needed the macros as below to get a successful cuda/10.1+xl/16.1.1 build:

```bash
__ibmxl_vrm__ -> __ibm__
```

**Review by ndellingwood — CHANGES_REQUESTED** [review]

File: core/src/Kokkos_Macros.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:157]: I can't find the documentation for `__ibm__`
I was looking AT https://www.ibm.com/support/knowledgecenter/SSGH3R_16.1.0/com.ibm.xlcpp161.aix.doc/compiler_ref/xlmacros.html

Comment (ndellingwood) [reply core/src/Kokkos_Macros.hpp]: Updated by commit 874d8c2

**Review by ndellingwood — APPROVED** [review]

File: core/src/Kokkos_Macros.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/Kokkos_Macros.hpp]: For the record `__ibmxl_vrm__` works as expected on Summit.

**Comment by dalg24** [comment]
Failures on Travis were timeouts.  Re-trigering.  AppVeyor failure is unrelated (#3596)

**Pull request merged**